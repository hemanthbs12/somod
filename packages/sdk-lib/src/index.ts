export * from "./tasks/packageJson";

export * from "./tasks/ignoreFiles";

export * from "./tasks/tsConfigBuildJson";

export * from "./tasks/init";

export * from "./tasks/build";

export * from "./tasks/link";

export * from "./tasks/dev";

export * from "./tasks/serverless";

export * from "./utils/constants";
