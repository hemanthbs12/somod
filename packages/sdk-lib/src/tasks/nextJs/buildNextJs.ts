import { childProcess } from "@sodaru/cli-base";

export const buildNextJs = async (dir: string): Promise<void> => {
  await childProcess(
    dir,
    process.platform === "win32" ? "npx.cmd" : "npx",
    ["next", "build"],
    { show: "on", return: "off" },
    { show: "on", return: "off" }
  );
};
