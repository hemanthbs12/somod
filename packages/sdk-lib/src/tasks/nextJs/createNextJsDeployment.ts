import {
  createNjpDeploymentPackageJson,
  copyNextDeployment
} from "../../utils/nextJs";

export const createNextJsDeployment = async (dir: string): Promise<void> => {
  await createNjpDeploymentPackageJson(dir);
  await copyNextDeployment(dir);
};
