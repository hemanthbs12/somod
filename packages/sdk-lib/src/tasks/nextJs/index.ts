export { watchRootModulePages } from "./watchRootModulePages";
export { watchRootModulePagesData } from "./watchRootModulePagesData";
export { watchRootModulePublicAssets } from "./watchRootModulePublicAssets";
export { nextCommand } from "./nextCommand";
export { vercelCommand } from "./vercelCommand";
export { createPages } from "./createPages";
export { createPublicAssets } from "./createPublicAssets";
export { validateUiConfigYaml } from "./validateUiConfigYaml";
export { buildUiConfigYaml } from "./buildUiConfigYaml";
export { generateNextConfig } from "./generateNextConfig";
export { buildUiPublic } from "./buildUiPublic";
export { validatePageExports } from "./validatePageExports";
export { validatePageData } from "./validatePageData";
export { deletePagesAndPublicDir } from "./deletePagesAndPublicDir";
