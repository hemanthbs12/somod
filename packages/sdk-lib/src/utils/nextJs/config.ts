import { ErrorSet, readJsonFileStore } from "@solib/cli-base";
import { existsSync } from "fs";
import { mkdir, writeFile } from "fs/promises";
import { isString, sortBy, uniqBy } from "lodash";
import { dirname, join } from "path";
import {
  file_configJson,
  file_configYaml,
  namespace_env_config,
  namespace_public_runtime_config,
  namespace_server_runtime_config,
  path_build,
  path_ui
} from "../constants";
import { freeze } from "../freeze";
import {
  JSONType,
  KeywordProcessor,
  KeywordValidator,
  parseJson,
  processKeywords,
  validateKeywords
} from "../jsonTemplate";
import { keywordAjvCompile } from "../keywords/ajv-compile";
import { keywordAnd } from "../keywords/and";
import { keywordEquals } from "../keywords/equals";
import { keywordIf } from "../keywords/if";
import { keywordJsonParse } from "../keywords/json-parse";
import { keywordJsonStringify } from "../keywords/json-stringify";
import { keywordKey } from "../keywords/key";
import { keywordOr } from "../keywords/or";
import { keywordParameter } from "../keywords/parameter";
import {
  KeywordDefinition,
  ModuleContent,
  ModuleContentMap
} from "../keywords/types";
import { Module, ModuleHandler, NamespaceLoader } from "../moduleHandler";
import { readYamlFileStore } from "../yamlFileStore";

const getBaseKeywords = () => [
  keywordAjvCompile,
  keywordAnd,
  keywordEquals,
  keywordIf,
  keywordJsonParse,
  keywordJsonStringify,
  keywordKey,
  keywordOr,
  keywordParameter
];

// this must match the @somod/ui-config-schema/schemas/index.json
export type Config = {
  env?: Record<string, JSONType>;
  imageDomains?: (string | JSONType)[];
  publicRuntimeConfig?: Record<string, JSONType>;
  serverRuntimeConfig?: Record<string, JSONType>;
};

export const loadConfig = async (module: Module): Promise<Config> => {
  const configPath = join(
    module.packageLocation,
    module.root ? "" : path_build,
    path_ui,
    module.root ? file_configYaml : file_configJson
  );
  if (existsSync(configPath)) {
    return module.root
      ? await readYamlFileStore(configPath)
      : await readJsonFileStore(configPath);
  } else {
    return {};
  }
};

export const loadConfigNamespaces: NamespaceLoader = async module => {
  const config = await loadConfig(module);
  return {
    [namespace_env_config]: Object.keys(config?.env || {}),
    [namespace_public_runtime_config]: Object.keys(
      config?.publicRuntimeConfig || {}
    ),
    [namespace_server_runtime_config]: Object.keys(
      config?.serverRuntimeConfig || {}
    )
  };
};

export const getModuleContentMap = async (
  modules: Module[]
): Promise<ModuleContentMap<Config>> => {
  const moduleContentMap: Record<string, ModuleContent<Config>> = {};

  await Promise.all(
    modules.map(async module => {
      const config = await loadConfig(module);
      moduleContentMap[module.name] = {
        moduleName: module.name,
        location: module.packageLocation,
        path: module.root
          ? `${path_ui}/${file_configYaml}`
          : `${path_build}/${path_ui}/${file_configJson}`,
        json: config
      };
    })
  );

  return freeze(moduleContentMap);
};

export const validate = async (
  dir: string,
  pluginKeywords: KeywordDefinition[] = []
) => {
  const moduleHandler = ModuleHandler.getModuleHandler();

  const moduleNodes = await moduleHandler.listModules();
  const moduleContentMap = await getModuleContentMap(
    moduleNodes.map(m => m.module)
  );
  const rootModuleName = (await moduleHandler.getRoodModuleNode()).module.name;

  const keywords = [...getBaseKeywords(), ...pluginKeywords];

  const keywordValidators: Record<string, KeywordValidator> = {};

  await Promise.all(
    keywords.map(async keyword => {
      const validator = await keyword.getValidator(
        dir,
        rootModuleName,
        moduleContentMap
      );

      keywordValidators[keyword.keyword] = validator;
    })
  );

  const errors = validateKeywords(
    parseJson(moduleContentMap[rootModuleName].json),
    keywordValidators
  );

  if (errors.length > 0) {
    throw new ErrorSet(errors);
  }
};

export const build = async (dir: string): Promise<void> => {
  const configYamlPath = join(dir, path_ui, file_configYaml);
  const yamlContentAsJson = (await readYamlFileStore(configYamlPath)) || {};

  const configJsonPath = join(dir, path_build, path_ui, file_configJson);

  await mkdir(dirname(configJsonPath), { recursive: true });
  await writeFile(configJsonPath, JSON.stringify(yamlContentAsJson));
};

export const generateCombinedConfig = async (
  dir: string,
  pluginKeywords: KeywordDefinition[] = []
): Promise<Config> => {
  const moduleHandler = ModuleHandler.getModuleHandler();
  const allModules = await moduleHandler.listModules();

  const moduleContentMap = await getModuleContentMap(
    allModules.map(m => m.module)
  );
  const rootModuleName = (await moduleHandler.getRoodModuleNode()).module.name;

  const keywords = [...getBaseKeywords(), ...pluginKeywords];

  const keywordProcessors: Record<string, KeywordProcessor> = {};

  await Promise.all(
    keywords.map(async keyword => {
      const processor = await keyword.getProcessor(
        dir,
        rootModuleName,
        moduleContentMap
      );

      keywordProcessors[keyword.keyword] = processor;
    })
  );

  const processsedMap: Record<string, Config> = {};

  allModules.reverse();
  allModules.forEach(moduleNode => {
    const moduleName = moduleNode.module.name;
    processsedMap[moduleName] = processKeywords(
      parseJson(moduleContentMap[moduleName].json),
      keywordProcessors
    ) as Config;
  });

  const namespaces = await moduleHandler.getNamespaces();

  const combinedImageDomains: Config["imageDomains"] = [];
  Object.values(processsedMap).forEach(config => {
    combinedImageDomains.push(...config.imageDomains);
  });

  const combinedEnv: Config["env"] = {};
  Object.keys(namespaces[namespace_env_config]).forEach(envName => {
    const moduleName = namespaces[namespace_env_config][envName];
    combinedEnv[envName] = processsedMap[moduleName].env[envName];
  });

  const combinedPublicRuntimeConfig: Config["publicRuntimeConfig"] = {};
  Object.keys(namespaces[namespace_public_runtime_config]).forEach(
    configName => {
      const moduleName =
        namespaces[namespace_public_runtime_config][configName];
      combinedPublicRuntimeConfig[configName] =
        processsedMap[moduleName].publicRuntimeConfig[configName];
    }
  );

  const combinedServerRuntimeConfig: Config["serverRuntimeConfig"] = {};
  Object.keys(namespaces[namespace_server_runtime_config]).forEach(
    configName => {
      const moduleName =
        namespaces[namespace_server_runtime_config][configName];
      combinedServerRuntimeConfig[configName] =
        processsedMap[moduleName].serverRuntimeConfig[configName];
    }
  );

  return {
    imageDomains: sortBy(
      uniqBy(combinedImageDomains, imageDomain =>
        isString(imageDomain) ? imageDomain : JSON.stringify(imageDomain)
      ),
      imageDomain =>
        isString(imageDomain) ? imageDomain : JSON.stringify(imageDomain)
    ),
    env: combinedEnv,
    publicRuntimeConfig: combinedPublicRuntimeConfig,
    serverRuntimeConfig: combinedServerRuntimeConfig
  };
};
