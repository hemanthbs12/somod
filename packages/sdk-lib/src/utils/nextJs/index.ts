import { copyDirectory, readJsonFileStore } from "@sodaru/cli-base";
import { existsSync } from "fs";
import { copyFile, mkdir, readdir, stat, writeFile } from "fs/promises";
import { join } from "path";
import {
  file_nextConfigJs,
  file_packageJson,
  file_packageLockJson,
  path_nextBuild,
  path_njp_deployment,
  path_njp_deployment_build,
  path_njp_working_dir,
  path_public
} from "../constants";

const getNjpDeploymentDir = (dir: string): string => {
  return join(dir, path_njp_working_dir, path_njp_deployment);
};

export const createNjpDeploymentPackageJson = async (dir: string) => {
  const njpDeploymentDir = getNjpDeploymentDir(dir);
  await mkdir(njpDeploymentDir, { recursive: true });

  const rootPackageLock = await readJsonFileStore(
    join(dir, file_packageLockJson)
  );

  await writeFile(
    join(njpDeploymentDir, file_packageJson),
    JSON.stringify(
      {
        name: rootPackageLock.name + "-deployment",
        version: rootPackageLock.version,
        scripts: {
          build: `mv ./${path_njp_deployment_build} ./${path_nextBuild}`
        },
        dependencies: {
          next: rootPackageLock.packages["node_modules/next"].version,
          react: rootPackageLock.packages["node_modules/react"].version,
          "react-dom":
            rootPackageLock.packages["node_modules/react-dom"].version
        }
      },
      null,
      2
    )
  );
};

export const copyNextDeployment = async (dir: string) => {
  const njpDeploymentDir = getNjpDeploymentDir(dir);
  const nextBuildContents = await readdir(join(dir, path_nextBuild));

  await mkdir(join(njpDeploymentDir, path_njp_deployment_build), {
    recursive: true
  });

  await Promise.all(
    nextBuildContents.map(async nextBuildContent => {
      if (nextBuildContent != "cache") {
        const sourcePath = join(dir, path_nextBuild, nextBuildContent);
        const targetPath = join(
          njpDeploymentDir,
          path_njp_deployment_build,
          nextBuildContent
        );
        const stats = await stat(sourcePath);
        if (stats.isDirectory()) {
          await copyDirectory(sourcePath, targetPath);
        } else {
          await copyFile(sourcePath, targetPath);
        }
      }
    })
  );
  if (existsSync(join(dir, file_nextConfigJs))) {
    await copyFile(
      join(dir, file_nextConfigJs),
      join(njpDeploymentDir, file_nextConfigJs)
    );
  }
  if (existsSync(join(dir, path_public))) {
    await copyDirectory(
      join(dir, path_public),
      join(njpDeploymentDir, path_public)
    );
  }
};
