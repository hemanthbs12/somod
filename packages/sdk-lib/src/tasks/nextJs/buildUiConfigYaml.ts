import { existsSync } from "fs";
import { join } from "path";
import { file_configYaml, path_ui } from "../../utils/constants";
import { buildConfig } from "../../utils/nextJs/config";

export const buildUiConfigYaml = async (
  dir: string,
  moduleIndicators: string[]
): Promise<void> => {
  const configYamlPath = join(dir, path_ui, file_configYaml);
  if (existsSync(configYamlPath)) {
    buildConfig(dir, moduleIndicators);
  }
};
