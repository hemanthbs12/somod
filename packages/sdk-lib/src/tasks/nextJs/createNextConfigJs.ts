import { writeFile } from "fs/promises";
import { join } from "path";
import { file_nextConfigJs, file_njpConfigJson } from "../../utils/constants";

export const createNextConfigJs = async (dir: string): Promise<void> => {
  await writeFile(
    join(dir, file_nextConfigJs),
    `/* eslint-disable */

  const fs = require("fs");
  const path = require("path");
  
  const njpConfigStr = fs.readFileSync(path.join(__dirname, "${file_njpConfigJson}"), {
    encoding: "utf8"
  });
  
  const njpConfig = JSON.parse(njpConfigStr);
  
  module.exports = njpConfig;
  `
  );
};
