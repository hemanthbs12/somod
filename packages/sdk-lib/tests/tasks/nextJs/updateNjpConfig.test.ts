import { createFiles, createTempDir, deleteDir } from "@sodev/test-utils";
import { existsSync } from "fs";
import { readFile } from "fs/promises";
import { join } from "path";
import { updateNjpConfig } from "../../../src";
import { Config } from "../../../src/utils/nextJs/config";
import { ErrorSet } from "@solib/cli-base";

describe("Test Task updateNjpConfig", () => {
  let dir: string = null;

  beforeEach(() => {
    dir = createTempDir();
  });

  afterEach(() => {
    deleteDir(dir);
  });

  test("with empty dir", async () => {
    createFiles(dir, {
      "package.json": JSON.stringify({ name: "sample", njp: "1.2.3" })
    });
    await expect(updateNjpConfig(dir, ["njp"])).resolves.toBeUndefined();
    await expect(
      readFile(join(dir, "njp.config.json"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
    await expect(
      readFile(join(dir, ".env"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
  });

  test("with content in root build dir", async () => {
    createFiles(dir, {
      "package.json": JSON.stringify({ name: "sample", njp: "1.2.3" }),
      "build/ui/config.json": JSON.stringify(
        {
          env: {
            DB_HOST: { default: "localhost", schema: { type: "string" } }
          },
          imageDomains: ["sodaru.com", "cloud.sodaru.com"],
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            }
          },
          serverRuntimeConfig: {
            googleRecaptchaApiKey: {
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      )
    });
    await expect(updateNjpConfig(dir, ["njp"])).resolves.toBeUndefined();
    await expect(
      readFile(join(dir, "njp.config.json"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
    await expect(
      readFile(join(dir, ".env"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
  });

  test("with validateOnly and content in root build dir", async () => {
    createFiles(dir, {
      "package.json": JSON.stringify({ name: "sample", njp: "1.2.3" }),
      "build/ui/config.json": JSON.stringify(
        {
          env: {
            DB_HOST: { default: "localhost", schema: { type: "string" } }
          },
          imageDomains: ["sodaru.com", "cloud.sodaru.com"],
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            }
          },
          serverRuntimeConfig: {
            googleRecaptchaApiKey: {
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      )
    });
    await expect(updateNjpConfig(dir, ["njp"], true)).resolves.toBeUndefined();
    expect(existsSync(join(dir, "njp.config.json"))).not.toBeTruthy();
    expect(existsSync(join(dir, ".env"))).not.toBeTruthy();
  });

  test("with conflicting content", async () => {
    createFiles(dir, {
      "package.json": JSON.stringify({
        name: "sample",
        njp: "1.2.3",
        dependencies: { pkg1: "^1.2.3", pkg2: "^2.3.4" }
      }),
      "build/ui/config.json": JSON.stringify({} as Config, null, 2),
      "node_modules/pkg1/package.json": JSON.stringify({
        name: "pkg1",
        njp: "1.2.3"
      }),
      "node_modules/pkg1/build/ui/config.json": JSON.stringify(
        {
          env: { DB_HOST: { schema: { type: "string" } } },
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            },
            recaptchaSiteKey: {
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      ),
      "node_modules/pkg2/package.json": JSON.stringify({
        name: "pkg2",
        njp: "1.2.3"
      }),
      "node_modules/pkg2/build/ui/config.json": JSON.stringify(
        {
          env: {
            DB_HOST: { schema: { type: "string" } },
            DB_PORT: { schema: { type: "string" } }
          },
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            }
          },
          serverRuntimeConfig: {
            recaptchaApiKey: {
              default: "",
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      )
    });
    await expect(updateNjpConfig(dir, ["njp"])).rejects.toEqual(
      new ErrorSet([
        new Error(
          "Error while resolving (pkg1, pkg2) modules for the env 'DB_HOST': Can not resolve"
        ),
        new Error(
          "Error while resolving (pkg1, pkg2) modules for the runtimeConfig 'theme': Can not resolve"
        )
      ])
    );
  });

  test("with resolved conflicting content", async () => {
    createFiles(dir, {
      "package.json": JSON.stringify({
        name: "sample",
        njp: "1.2.3",
        dependencies: { pkg1: "^1.2.3", pkg2: "^2.3.4" }
      }),
      "build/ui/config.json": JSON.stringify(
        {
          env: {
            DB_HOST: { schema: { type: "string" }, default: "localhost" },
            DB_PORT: { schema: { type: "string" }, default: "8080" }
          },
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#000000" } } },
              label: "Theme",
              schema: { type: "object" }
            }
          }
        } as Config,
        null,
        2
      ),
      "node_modules/pkg1/package.json": JSON.stringify({
        name: "pkg1",
        njp: "1.2.3"
      }),
      "node_modules/pkg1/build/ui/config.json": JSON.stringify(
        {
          env: { DB_HOST: { schema: { type: "string" } } },
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            },
            recaptchaSiteKey: {
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      ),
      "node_modules/pkg2/package.json": JSON.stringify({
        name: "pkg2",
        njp: "1.2.3"
      }),
      "node_modules/pkg2/build/ui/config.json": JSON.stringify(
        {
          env: {
            DB_HOST: { schema: { type: "string" } },
            DB_PORT: { schema: { type: "string" } }
          },
          runtimeConfig: {
            theme: {
              default: { palette: { primary: { main: "#004b89" } } },
              label: "Theme",
              schema: { type: "object" }
            }
          },
          serverRuntimeConfig: {
            recaptchaApiKey: {
              default: "",
              schema: { type: "string" }
            }
          }
        } as Config,
        null,
        2
      )
    });
    await expect(updateNjpConfig(dir, ["njp"])).resolves.toBeUndefined();
    await expect(
      readFile(join(dir, "njp.config.json"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
    await expect(
      readFile(join(dir, ".env"), { encoding: "utf8" })
    ).resolves.toMatchSnapshot();
  });
});
