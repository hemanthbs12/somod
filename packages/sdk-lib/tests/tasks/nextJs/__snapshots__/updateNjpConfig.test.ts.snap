// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Task updateNjpConfig with content in root build dir 1`] = `
"{
  \\"images\\": {
    \\"domains\\": [
      \\"sodaru.com\\",
      \\"cloud.sodaru.com\\"
    ]
  },
  \\"publicRuntimeConfig\\": {
    \\"theme\\": {
      \\"palette\\": {
        \\"primary\\": {
          \\"main\\": \\"#004b89\\"
        }
      }
    }
  },
  \\"serverRuntimeConfig\\": {}
}"
`;

exports[`Test Task updateNjpConfig with content in root build dir 2`] = `"DB_HOST=localhost"`;

exports[`Test Task updateNjpConfig with empty dir 1`] = `
"{
  \\"images\\": {
    \\"domains\\": []
  },
  \\"publicRuntimeConfig\\": {},
  \\"serverRuntimeConfig\\": {}
}"
`;

exports[`Test Task updateNjpConfig with empty dir 2`] = `""`;

exports[`Test Task updateNjpConfig with resolved conflicting content 1`] = `
"{
  \\"images\\": {
    \\"domains\\": []
  },
  \\"publicRuntimeConfig\\": {
    \\"theme\\": {
      \\"palette\\": {
        \\"primary\\": {
          \\"main\\": \\"#000000\\"
        }
      }
    }
  },
  \\"serverRuntimeConfig\\": {
    \\"recaptchaApiKey\\": \\"\\"
  }
}"
`;

exports[`Test Task updateNjpConfig with resolved conflicting content 2`] = `
"DB_HOST=localhost
DB_PORT=8080"
`;
